File: rtiow/Rust/main04.rs
use std::ops::Div;
use std::ops::Mul;

#[derive(Debug)]
#[derive(Copy, Clone)]
struct Vec3([f32; 3]);

#[derive(Debug)]
struct Ray {
	origin: Vec3,
	direction: Vec3
}

impl Vec3 {
	fn squared_length(self) -> f32 {
		let Self([x1, y1, z1]) = self;
		x1 * x1 + y1 * y1 + z1 * z1
	}
	fn length(self) -> f32 {
		f32::sqrt(self.squared_length())
	}
}

impl std::ops::Div for Vec3 {
    type Output = Vec3;

    fn div(self, rhs: Vec3) -> Self::Output {
        let Self([x1, y1, z1]) = self;
        let Self([x2, y2, z2]) = rhs;

        Self([x1 / x2, y1 / y2, z1 / z2])
    }
}

impl std::ops::Add for Vec3 {
    type Output = Vec3;

    fn add(self, rhs: Vec3) -> Self::Output {
        let Self([x1, y1, z1]) = self;
        let Self([x2, y2, z2]) = rhs;

        Self([x1 + x2, y1 + y2, z1 + z2])
    }
}

impl Div<f32> for Vec3 {
	type Output = Self;

	fn div(self, rhs: f32) -> Self::Output {
		let Self([x1, y1, z1]) = self;
		Self([x1 / rhs, y1 / rhs, z1 / rhs])
	}
}

impl Mul<f32> for Vec3 {
	type Output = Self;

	fn mul(self, rhs: f32) -> Self::Output {
		let Self([x1, y1, z1]) = self;
		Self([x1 * rhs, y1 * rhs, z1 * rhs])
	}
}

//let v = Vec3([0., 0., 0.]);
//v.0[1]

fn unit_vector(v: Vec3) -> Vec3 {
	v / v.length()
}

fn color(r: Ray) -> Vec3 {
//	println!("r={:?}", r);
	let unit_direction = unit_vector(r.direction);
//	println!("ud={:?}", unit_direction);
	let t = 0.5 * (unit_direction.0[1] + 1.);
//	println!("t={}", t);
	Vec3([1., 1., 1.]) * (1. - t) + Vec3([0.5, 0.7, 1.]) * t
}

fn main() {
	let nx = 200;
	let ny = 100;
	println!("P3");
	println!("{} {}", nx, ny);
	println!("255");
	let lower_left_corner = Vec3([-2.0, -1.0, -1.0]);
	let horizontal = Vec3([4., 0., 0.]);
	let vertical = Vec3([0., 2., 0.]);
	let origin = Vec3([0., 0., 0.]);
	for j in (0..ny).rev() {
		for i in 0..nx {
			let u = i as f32 / nx as f32;
			let v = j as f32 / ny as f32;
//			println!("u={} v={}", u, v);
			let r = Ray {
				origin: origin,
				direction: lower_left_corner + horizontal * u + vertical * v
			};
			let col = color(r);
//			println!("col={:?}", col);
			let ir = (255.99*col.0[0]) as i32;
			let ig = (255.99*col.0[1]) as i32;
			let ib = (255.99*col.0[2]) as i32;
			println!("{} {} {}", ir, ig, ib);
		}
	}
}

File: rtiow/C/main04.c
#include <stdio.h>
#include <stdlib.h>

#include "ray.h"
#include "vec3.h"

void color(vec3 col, const ray r) {
  vec3 unit_direction;
  unit_vector(unit_direction, r.direction);
  //   printf("unit_direction: %f,%f,%f\n", unit_direction[0],
  //   unit_direction[1],
  //          unit_direction[2]);
  float t = 0.5 * (unit_direction[1] + 1.0);
  vec3 col0 = {1.0, 1.0, 1.0};
  vec3 col1 = {0.5, 0.7, 1.0};
  vmul(col0, 1.0 - t, col0);
  vmul(col1, t, col1);
  vadd(col, col0, col1);
}

int main() {
  int nx = 200;
  int ny = 100;
  printf("P3\n");
  printf("%d %d\n", nx, ny);
  printf("255\n");
  vec3 lower_left_corner = {-2.0, -1.0, -1.0};
  //	vprint(lower_left_corner);
  vec3 horizontal = {4.0, 0.0, 0.0};
  vec3 vertical = {0.0, 2.0, 0.0};
  vec3 origin = {0.0, 0.0, 0.0};
  for (int j = ny - 1; j >= 0; j--) {
    for (int i = 0; i < nx; i++) {
      float u = (float)i / (float)nx;
      float v = (float)j / (float)ny;
      //   printf("j=%d i=%d u=%f v=%f\n", j, i, u, v);
      ray r;
      vec3 direction, direction0, direction1;
      vmul(direction0, v, vertical);
      vmul(direction1, u, horizontal);
      vadd(direction, direction0, direction1);
      vadd(direction, direction, lower_left_corner);
      rmake(&r, origin, direction);
      //			rprint(&r);exit(0);
      vec3 col;
      color(col, r);
      //   printf("%f %f %f\n", col[0], col[1], col[2]);
      int ir = (int)(255.99f * col[0]);
      int ig = (int)(255.99f * col[1]);
      int ib = (int)(255.99f * col[2]);
      printf("%d %d %d\n", ir, ig, ib);
    }
  }
  return 0;
}
